src/python/pants/engine/isolated_process.py

```diff
@@ -153,9 +164,105 @@ def create(product_type, binary_type, input_selectors, input_conversion, output_
     # Return a task triple that executes the function to produce the product type.
     return TaskRule(product_type, inputs, func)
 
-
 def create_snapshot_rules():
   """Intrinsically replaced on the rust side."""
   return [
       SingletonRule(_Snapshots, _Snapshots('/dev/null'))
     ]
+
+def _execute_process(input_conversion,
+  output_conversion,
+  snapshot_directory,
+  binary,
+  *args):
+  """A pickleable top-level function to execute a process.
+
+  Receives two conversion functions, some required inputs, and the user-declared inputs.
+  """
+
+  process_request = input_conversion(*args)
+  _setup_process_execution(process_request, snapshot_directory)
+
+  # TODO resolve what to do with output files, then make these tmp dirs cleaned up.
+  with temporary_dir(cleanup=False) as sandbox_dir:
+    if process_request.snapshots:
+      for snapshot in process_request.snapshots:
+        _extract_snapshot(snapshot_directory.root, snapshot, sandbox_dir)
+
+    # All of the snapshots have been checked out now.
+    if process_request.directories_to_create:
+      for d in process_request.directories_to_create:
+        safe_mkdir(os.path.join(sandbox_dir, d))
+
+    command = binary.prefix_of_command() + tuple(process_request.args)
+    logger.debug('Running command: "{}" in {}'.format(command, sandbox_dir))
+    # TODO At some point, we may want to replace this blocking wait with a timed one that returns
+    # some kind of in progress state.
+    req = ExecuteProcessRequest(command, dict(os.environ))
+
+    return _post_process_execution(output_conversion, snapshot_directory)
+
+class ExecuteProcess(object):
+  """A static helper for defining a task rule to execute a process."""
+
+  def __new__(cls, *args):
+    raise ValueError('Use `create` to declare a task function representing a process.')
+
+  @staticmethod
+  def create_in(product_type, input_selectors, input_conversion):
+    # TODO: combine create_in/create_out fucntions
+    func = functools.partial(_setup_process_execution, input_conversion)
+    func.__name__ = '{}_and_then_execute_process'.format(input_conversion.__name__)
+    inputs = [Select(_Snapshots)] + list(input_selectors)
+
+    # Return a task triple that executes the function to produce the product type.
+    return TaskRule(product_type, inputs, func)
+
+  @staticmethod
+  def create_out(product_type, input_selectors, output_conversion):
+    func = functools.partial(_post_process_execution,
+      output_conversion)
+    func.__name__ = 'execute_process_and_then_{}'.format(output_conversion.__name__)
+    inputs = list(input_selectors)
+
+    # Return a task triple that executes the function to produce the product type.
+    return TaskRule(product_type, inputs, func)
+
+
+class ExecuteProcessRequest(datatype('ExecuteProcessRequest', ['args', 'env'])):
+  """Request for execution with args and snapshots to extract."""
+
+  def __new__(cls, args, env, **kwargs):
```

Comment:
Does this have to take kwargs? If so, can you document why? it's a little strange to pass them through, given I *think* if you pass any they'll cause an error?